pkg/cgroups/cgroupblkio.go

```diff
@@ -50,51 +52,99 @@ var blkioThrottleWriteIOPSFiles = []string{"blkio.throttle.write_iops_device"}
 //  other |  Write to cgroups, sets the value
 type OciBlockIOParameters struct {
 	Weight                  int64
-	WeightDevice            []OciWeightDeviceParameters
-	ThrottleReadBpsDevice   []OciRateDeviceParameters
-	ThrottleWriteBpsDevice  []OciRateDeviceParameters
-	ThrottleReadIOPSDevice  []OciRateDeviceParameters
-	ThrottleWriteIOPSDevice []OciRateDeviceParameters
+	WeightDevice            OciDeviceWeights
+	ThrottleReadBpsDevice   OciDeviceRates
+	ThrottleWriteBpsDevice  OciDeviceRates
+	ThrottleReadIOPSDevice  OciDeviceRates
+	ThrottleWriteIOPSDevice OciDeviceRates
 }
 
-// OciWeightDeviceParameters contains values for
+// OciDeviceWeight contains values for
 // - blkio.[io-scheduler].weight
-type OciWeightDeviceParameters struct {
+type OciDeviceWeight struct {
 	Major  int64
 	Minor  int64
 	Weight int64
 }
 
-// OciRateDeviceParameters contains values for
+// OciDeviceRate contains values for
 // - blkio.throttle.read_bps_device
 // - blkio.throttle.write_bps_device
 // - blkio.throttle.read_iops_device
 // - blkio.throttle.write_iops_device
-type OciRateDeviceParameters struct {
+type OciDeviceRate struct {
 	Major int64
 	Minor int64
 	Rate  int64
 }
 
+// OciDeviceWeights contains weights for devices
+type OciDeviceWeights []OciDeviceWeight
+
+// OciDeviceRates contains throttling rates for devices
+type OciDeviceRates []OciDeviceRate
+
+// OciDeviceParameters interface provides functions common to OciDeviceWeights and OciDeviceRates
+type OciDeviceParameters interface {
+	Append(maj, min, val int64)
+	Update(maj, min, val int64)
+}
+
+// Append appends (major, minor, value) to OciDeviceWeights slice.
+func (w *OciDeviceWeights) Append(maj, min, val int64) {
+	*w = append(*w, OciDeviceWeight{Major: maj, Minor: min, Weight: val})
+}
+
+// Append appends (major, minor, value) to OciDeviceRates slice.
+func (r *OciDeviceRates) Append(maj, min, val int64) {
+	*r = append(*r, OciDeviceRate{Major: maj, Minor: min, Rate: val})
+}
+
+// Update updates device weight in OciDeviceWeights slice, or appends it if not found.
+func (w *OciDeviceWeights) Update(maj, min, val int64) {
+	if w != nil {
+		for index, devWeight := range *w {
+			if devWeight.Major == maj && devWeight.Minor == min {
+				(*w)[index].Weight = val
+				return
+			}
+		}
+	}
+	w.Append(maj, min, val)
```

Comment:
This will segfault if w is nil.

Do we even want to check for `w != nil`? And thus, silently ignore some possible bugs.

Reply:
Good finding, thanks!

I removed the unnecessary `w != nil` check. Now if `.Append` or `.Update` are called with `nil` slice pointer, it'll cause a runtime error in the `for ... range *list` loop.